File: src/test/java/ai/ancf/lmos/operator/reconciler/ChannelReconcilerTest.java
/*
 * SPDX-FileCopyrightText: 2024 Deutsche Telekom AG
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package ai.ancf.lmos.operator.reconciler;

import io.fabric8.kubernetes.client.KubernetesClient;
import io.fabric8.kubernetes.client.utils.Serialization;
import io.javaoperatorsdk.operator.springboot.starter.test.EnableMockOperator;
import ai.ancf.lmos.operator.resources.agent.AgentResource;
import ai.ancf.lmos.operator.resources.channel.ChannelResource;
import ai.ancf.lmos.operator.resources.channel.ResolveStatus;
import ai.ancf.lmos.operator.resources.routing.ChannelRoutingResource;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.util.ResourceUtils;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.Assertions.assertThat;
import static org.awaitility.Awaitility.await;

@SpringBootTest
@EnableMockOperator(crdPaths = {"classpath:META-INF/fabric8/channels.lmos.ai-v1.yml",
        "classpath:META-INF/fabric8/agents.lmos.ai-v1.yml",
        "classpath:META-INF/fabric8/channelroutings.lmos.ai-v1.yml",
        "classpath:META-INF/fabric8/channelrollouts.lmos.ai-v1.yml",
        })
class ChannelReconcilerTest {
    @Autowired
    KubernetesClient client;

    @AfterEach
    public void cleanUp() {
        client.resources(AgentResource.class).delete();
        client.resources(ChannelResource.class).delete();
        client.resources(ChannelRoutingResource.class).delete();
    }

    @Test
    void verifyThatCRDAreCreated() {

        assertThat(
                client
                        .apiextensions()
                        .v1()
                        .customResourceDefinitions()
                        .withName("agents.lmos.ai")
                        .get())
                .isNotNull();

        assertThat(
                client
                        .apiextensions()
                        .v1()
                        .customResourceDefinitions()
                        .withName("channels.lmos.ai")
                        .get())
                .isNotNull();
    }

    @Test
    void shouldCreateResolvedChannelRouting() throws FileNotFoundException {

        // When I create two Agents
        client.load(getResource("billing-agent-v1.yaml")).createOrReplace();
        client.load(getResource("contract-agent-v1.yaml")).createOrReplace();

        // When I create a Channel the Reconciler should start
        client.load(getResource("ivr-channel-v1.yaml")).createOrReplace();

        // Then the channel status should be updated to resolved
        var channelResources = await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(() -> client.resources(ChannelResource.class).list().getItems(), c -> c.get(0).getStatus() != null);

        assertThat(channelResources).isNotNull().hasSize(1);
        assertThat(channelResources.get(0).getStatus().getResolveStatus()).isEqualTo(ResolveStatus.RESOLVED);
        assertThat(channelResources.get(0).getStatus().getUnresolvedRequiredCapabilities()).isNull();

        // And the channel routing resource should be created and status is resolved
        var channelRoutingResources = await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(() -> client.resources(ChannelRoutingResource.class).list().getItems(), r -> !r.isEmpty());

        assertThat(channelRoutingResources).hasSize(1);
        ChannelRoutingResource channelRoutingResource = channelRoutingResources.get(0);
        //assertThat(channelRoutingResource.getStatus().getResolveStatus()).isEqualTo(ResolveStatus.RESOLVED);
        assertThat(channelRoutingResource.getMetadata().getName()).isEqualTo("ivr-stable");
        assertThat(channelRoutingResource.getMetadata().getOwnerReferences()).hasSize(1);
        System.out.println(Serialization.asYaml(channelRoutingResource));

        // If I delete the ChannelRoutingResource the desired state is reconciled again
        client.resources(ChannelRoutingResource.class).delete();
        channelRoutingResources = await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(() -> client.resources(ChannelRoutingResource.class).list().getItems(), r -> !r.isEmpty());

        assertThat(channelRoutingResources).hasSize(1);
    }

    @Test
    void shouldCreateUnresolvedChannelRouting() throws FileNotFoundException, InterruptedException {

        // When I create an Agent and a Channel
        client.load(getResource("billing-agent-v1.yaml")).createOrReplace();
        client.load(getResource("web-channel-v1.yaml")).createOrReplace();

        // Then the channel status should be updated to unresolved
        var channelResources = await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(() -> client.resources(ChannelResource.class).list().getItems(), c -> c.get(0).getStatus() != null);

        assertThat(channelResources).isNotNull().hasSize(1);
        assertThat(channelResources.get(0).getStatus().getResolveStatus()).isEqualTo(ResolveStatus.UNRESOLVED);
        assertThat(channelResources.get(0).getStatus().getUnresolvedRequiredCapabilities()).isNotEmpty();

        // And the channel routing resource should be created but status is unresolved
        var channelRoutingResources = await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(() -> client.resources(ChannelRoutingResource.class).list().getItems(), r -> !r.isEmpty());

        assertThat(channelRoutingResources).isNotNull().hasSize(1);
        ChannelRoutingResource channelRoutingResource = channelRoutingResources.get(0);
        //assertThat(channelRoutingResource.getStatus().getResolveStatus()).isEqualTo(ResolveStatus.UNRESOLVED);
        assertThat(channelRoutingResource.getMetadata().getName()).isEqualTo("web-stable");
        assertThat(channelRoutingResource.getMetadata().getOwnerReferences()).hasSize(1);
    }

    @NotNull
    private static FileInputStream getResource(String resourceName) throws FileNotFoundException {
        return new FileInputStream(ResourceUtils.getFile("classpath:" + resourceName));
    }

}

File: src/test/kotlin/org/eclipse/lmos/operator/reconciler/ChannelReconcilerTest.kt
/*
 * SPDX-FileCopyrightText: 2024 Deutsche Telekom AG
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.eclipse.lmos.operator.reconciler

import io.fabric8.kubernetes.client.KubernetesClient
import io.fabric8.kubernetes.client.utils.Serialization
import io.javaoperatorsdk.operator.springboot.starter.test.EnableMockOperator
import org.assertj.core.api.Assertions.assertThat
import org.awaitility.Awaitility
import org.eclipse.lmos.operator.OperatorApplication
import org.eclipse.lmos.operator.resources.AgentResource
import org.eclipse.lmos.operator.resources.ChannelResource
import org.eclipse.lmos.operator.resources.ChannelRoutingResource
import org.eclipse.lmos.operator.resources.ResolveStatus
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.util.ResourceUtils
import java.io.FileInputStream
import java.util.concurrent.TimeUnit

@Disabled
@SpringBootTest(
    properties = ["spring.main.allow-bean-definition-overriding=true"],
    classes = [OperatorApplication::class],
)
@EnableMockOperator(
    crdPaths = [
        "classpath:META-INF/fabric8/channels.lmos.eclipse-v1.yml",
        "classpath:META-INF/fabric8/agents.lmos.eclipse-v1.yml",
        "classpath:META-INF/fabric8/channelroutings.lmos.eclipse-v1.yml",
        "classpath:META-INF/fabric8/channelrollouts.lmos.eclipse-v1.yml",
    ],
)
internal class ChannelReconcilerTest {
    @Autowired
    lateinit var client: KubernetesClient

    @AfterEach
    fun cleanUp() {
        client.resources(AgentResource::class.java).delete()
        client.resources(ChannelResource::class.java).delete()
        client.resources(ChannelRoutingResource::class.java).delete()
    }

    @Test
    fun verifyThatCRDAreCreated() {
        assertThat(
            client
                .apiextensions()
                .v1()
                .customResourceDefinitions()
                .withName("agents.lmos.eclipse")
                .get(),
        )
            .isNotNull()

        assertThat(
            client
                .apiextensions()
                .v1()
                .customResourceDefinitions()
                .withName("channels.lmos.eclipse")
                .get(),
        )
            .isNotNull()
    }

    @Test
    fun shouldCreateResolvedChannelRouting() {
        // When I create two Agents

        client.load(getResource("acme-billing-agent-v1.yaml")).createOrReplace()
        client.load(getResource("contract-agent-v1.yaml")).createOrReplace()

        // When I create a Channel the Reconciler should start
        client.load(getResource("acme-ivr-channel-v1.yaml")).createOrReplace()

        // Then the channel status should be updated to resolved
        val channelResources =
            Awaitility.await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(
                    {
                        client.resources(
                            ChannelResource::class.java,
                        ).list().items
                    },
                    { c: List<ChannelResource> -> c[0].status != null },
                )

        assertThat(channelResources).isNotNull().hasSize(1)
        assertThat(channelResources[0].status.resolveStatus).isEqualTo(ResolveStatus.RESOLVED)
        assertThat(channelResources[0].status.unresolvedRequiredCapabilities).isEmpty()

        // And the channel routing resource should be created and status is resolved
        var channelRoutingResources =
            Awaitility.await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(
                    {
                        client.resources(
                            ChannelRoutingResource::class.java,
                        ).list().items
                    },
                    { r: List<ChannelRoutingResource> -> r.isNotEmpty() },
                )

        assertThat(channelRoutingResources).hasSize(1)
        val channelRoutingResource = channelRoutingResources[0]
        // assertThat(channelRoutingResource.getStatus().getResolveStatus()).isEqualTo(ResolveStatus.RESOLVED);
        assertThat(channelRoutingResource.metadata.name).isEqualTo("acme-ivr-stable")
        assertThat(channelRoutingResource.metadata.ownerReferences).hasSize(1)
        println(Serialization.asYaml(channelRoutingResource))

        // If I delete the ChannelRoutingResource the desired state is reconciled again
        client.resources(ChannelRoutingResource::class.java).delete()
        channelRoutingResources =
            Awaitility.await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(
                    {
                        client.resources(
                            ChannelRoutingResource::class.java,
                        ).list().items
                    },
                    { r: List<ChannelRoutingResource> -> r.isNotEmpty() },
                )

        assertThat(channelRoutingResources).hasSize(1)
    }

    @Test
    fun shouldCreateUnresolvedChannelRouting() {
        // When I create an Agent and a Channel

        client.load(getResource("acme-billing-agent-v1.yaml")).createOrReplace()
        client.load(getResource("acme-web-channel-v1.yaml")).createOrReplace()

        // Then the channel status should be updated to unresolved
        val channelResources =
            Awaitility.await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(
                    {
                        client.resources(
                            ChannelResource::class.java,
                        ).list().items
                    },
                    { c: List<ChannelResource> -> c[0].status != null },
                )

        assertThat(channelResources).isNotNull().hasSize(1)
        assertThat(channelResources[0].status.resolveStatus).isEqualTo(ResolveStatus.UNRESOLVED)
        assertThat(channelResources[0].status.unresolvedRequiredCapabilities).isNotEmpty()

        // And the channel routing resource should be created but status is unresolved
        val channelRoutingResources =
            Awaitility.await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(
                    {
                        client.resources(
                            ChannelRoutingResource::class.java,
                        ).list().items
                    },
                    { r: List<ChannelRoutingResource> -> r.isNotEmpty() },
                )

        assertThat(channelRoutingResources).isNotNull().hasSize(1)
        val channelRoutingResource = channelRoutingResources[0]
        // assertThat(channelRoutingResource.getStatus().getResolveStatus()).isEqualTo(ResolveStatus.UNRESOLVED);
        assertThat(channelRoutingResource.metadata.name).isEqualTo("acme-web-stable")
        assertThat(channelRoutingResource.metadata.ownerReferences).hasSize(1)
    }

    @Test
    fun shouldCreateUnresolvedChannelRoutingForNotMatchingTenant() {
        // When I create an Agent and a Channel with different tenants

        client.load(getResource("acme-billing-agent-v1.yaml")).createOrReplace()
        client.load(getResource("globex-web-channel-v1.yaml")).createOrReplace()

        // Then the channel status should be updated to unresolved
        val channelResources =
            Awaitility.await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(
                    {
                        client.resources(
                            ChannelResource::class.java,
                        ).list().items
                    },
                    { c: List<ChannelResource> -> c[0].status != null },
                )

        assertThat(channelResources).isNotNull().hasSize(1)
        assertThat(channelResources[0].status.resolveStatus).isEqualTo(ResolveStatus.UNRESOLVED)
        assertThat(channelResources[0].status.unresolvedRequiredCapabilities).isNotEmpty()

        // And the channel routing resource should be created but status is unresolved
        val channelRoutingResources =
            Awaitility.await().atMost(5, TimeUnit.SECONDS) // Timeout duration
                .pollInterval(50, TimeUnit.MILLISECONDS) // Polling interval
                .until(
                    {
                        client.resources(
                            ChannelRoutingResource::class.java,
                        ).list().items
                    },
                    { r: List<ChannelRoutingResource> -> r.isNotEmpty() },
                )

        assertThat(channelRoutingResources).isNotNull().hasSize(1)
        val channelRoutingResource = channelRoutingResources[0]
        // assertThat(channelRoutingResource.getStatus().getResolveStatus()).isEqualTo(ResolveStatus.UNRESOLVED);
        assertThat(channelRoutingResource.metadata.name).isEqualTo("globex-web-stable")
        assertThat(channelRoutingResource.metadata.ownerReferences).hasSize(1)
    }

    private fun getResource(resourceName: String): FileInputStream {
        return FileInputStream(ResourceUtils.getFile("classpath:$resourceName"))
    }
}
